feat(mini.files): add cwd and vertical/horizontal keybindings to mini.files

dusty-phillips opened the pull request:
* Makes the toggle_hidden keybinding configurable via mappings
* Adds new mini.files keybindings for opening files in a vertical split, with both go_in and go_in_plus modes (configurable via mappings)
* Adds new keybinding to change working directory from mini.files (configurable via mappings)

Closes #2692

dusty-phillips commented on lua/lazyvim/plugins/extras/editor/mini-files.lua:
```diff
@@ -32,7 +32,7 @@ return {
 
     local show_dotfiles = true
-    local filter_show = function(fs_entry)
+    local filter_show = function(_)
```
stylelua was warning me about unused variable here.

dusty-phillips commented on lua/lazyvim/plugins/extras/editor/mini-files.lua:
```diff
@@ -45,12 +45,69 @@ return {
       require("mini.files").refresh({ content = { filter = new_filter } })
     end
 
+    local map_split = function(buf_id, lhs, direction, close_on_file)
```
This function is modified from the example in the mini.files repo quite a bit. notably, I made it open the selected file after creating the split, and added close_on_file so that we can have `_plus` keybindings. I also made the keybindings configurable via opts.mappings.

dusty-phillips commented on lua/lazyvim/plugins/extras/editor/mini-files.lua:
```diff
@@ -45,12 +45,69 @@ return {
+          "n",
+          opts.mappings.toggle_hidden or "g.",
+          toggle_dotfiles,
+          { buffer = buf_id, desc = "Toggle hidden files" }
```
Added description here, though I also have a separate PR open for that in case you don't want to accept these more invasive changes.

dusty-phillips commented on lua/lazyvim/plugins/extras/editor/mini-files.lua:
```diff
@@ -45,12 +45,69 @@ return {
-        vim.keymap.set("n", "g.", toggle_dotfiles, { buffer = buf_id })
+        vim.keymap.set(
+          "n",
+          opts.mappings.toggle_hidden or "g.",
```
Added ability to configure this existing mapping from `opts`

dusty-phillips commented on lua/lazyvim/plugins/extras/editor/mini-files.lua:
```diff
@@ -45,12 +45,69 @@ return {
+        map_split(buf_id, opts.mappings.go_in_horizontal or "gs", "horizontal", false)
+        map_split(buf_id, opts.mappings.go_in_vertical or "gv", "vertical", false)
+        map_split(buf_id, opts.mappings.go_in_horizontal_plus or "gS", "horizontal", true)
+        map_split(buf_id, opts.mappings.go_in_vertical_plus or "gV", "vertical", true)
```
The first two mappings (as well as `gc` below) were suggested by mini.files documentation. I don't care for them as they conflict with existing mappings. Happy to change them to something more Lazy, but I'm not sure what they should be. Some ideas:

* `<space>|` and `<space><minus>`, overwriting the LazyVim keymappings for opening new windows in splits, but not obvious what to do with the `*_plus` versions.
*  `<C-w>v, <C-w>V, `<C-w>s` and `<C-w>S`, overwriting the standard vim keymappings for opening new windows in splits.
* `\v`, `\V`, `\s`, `\S` which is what I'll use because I have `\` mapped to `<C-w>`

pkazmier commented:
I'm very excited about these proposed enhancements for the mini.files extra. I've recently discovered mini.files, which has allowed me to completely disable neo-tree.

Re: which bindings to use, I'd vote for C-w v,V,s,S as pressing those bindings while in mini.files windows doesn't make sense and does weird stuff to the screen anyways.

briandipalma commented:
`s` is used to search. I would use it to move around quickly even in mini.files.

dusty-phillips commented:
* Rebased and fixed conflicts.
* Switched to `<ctrl-w>s` and `<ctrl-w>v` for opening in splits

Tangentially related: my book recommends remapping `\` to `ctrl-w`, and I suggest this could be a default in LazyVim. Then we don't need space-| and space-<minus> anymore because \v and \s are just as easy to type.

dpetka2001 commented:
Kinda naive question, but is there a specific reason we use 2 separate autocmds on the same event?

dusty-phillips commented:
> Kinda naive question, but is there a specific reason we use 2 separate autocmds on the same event?

It was cause in my local config I wanted to keep my changes separate from the LazyVim configuration, but since this IS the (proposed) LazyVim configuration it doesn't make sense here. I think I pulled a folke and pushed a change for it while you were asking the question. 😂

dusty-phillips commented:
I think that's just me being daft.

krishna-bala commented:
Hi @dusty-phillips,
I'm getting an error from this PR:
   Error  12:55:46 PM msg_show.lua_error Error detected while processing User Autocommands for "MiniFilesBufferCreate":
12:55:46 PM msg_show Error executing lua callback: ...LazyVim/lua/lazyvim/plugins/extras/editor/mini-files.lua:85: attempt to index field 'mappings' (a nil value)
stack traceback:
	...LazyVim/lua/lazyvim/plugins/extras/editor/mini-files.lua:85: in function <...LazyVim/lua/lazyvim/plugins/extras/editor/mini-files.lua:80>
	[C]: in function 'nvim_exec_autocmds'
	...hna/.local/share/nvim/lazy/mini.files/lua/mini/files.lua:2619: in function 'trigger_event'
	...hna/.local/share/nvim/lazy/mini.files/lua/mini/files.lua:1890: in function 'buffer_create'
	...hna/.local/share/nvim/lazy/mini.files/lua/mini/files.lua:1756: in function 'view_ensure_proper'
	...hna/.local/share/nvim/lazy/mini.files/lua/mini/files.lua:1529: in function 'explorer_refresh_depth_window'
	...hna/.local/share/nvim/lazy/mini.files/lua/mini/files.lua:1321: in function 'explorer_refresh'
	...hna/.local/share/nvim/lazy/mini.files/lua/mini/files.lua:714: in function 'open'
	...LazyVim/lua/lazyvim/plugins/extras/editor/mini-files.lua:19: in function <...LazyVim/lua/lazyvim/plugins/extras/editor/mini-files.lua:18>
Reverting to d704ca4c1b761303c0c1119341a566d161ab12a3, I do not see these errors. Seems to be upset about `opts.mappings`. I can dive a bit deeper into why this is happening, but I wanted to post now for awareness.
(posted this message initially on the commit)

dpetka2001 commented:
@krishna-bala If you'd like you can check out #3507 to see if it works correctly. Hopefully I didn't miss anything.
